Custom configuration classes break numpydoc

### Describe the bug

#13783 changes the type of the configuration passed to extensions from (I assume) a dict to a custom class.  numpy doc has code that looks like

```python
def mangle_docstrings(app: SphinxApp, what, name, obj, options, lines):
    if DEDUPLICATION_TAG in lines:
        return
    show_inherited_class_members = app.config.numpydoc_show_inherited_class_members
    if isinstance(show_inherited_class_members, dict):
        try:
            show_inherited_class_members = show_inherited_class_members[name]
        except KeyError:
            show_inherited_class_members = True

    cfg = {
        "use_plots": app.config.numpydoc_use_plots,
        "show_class_members": app.config.numpydoc_show_class_members,
        "show_inherited_class_members": show_inherited_class_members,
        "class_members_toctree": app.config.numpydoc_class_members_toctree,
        "attributes_as_param_list": app.config.numpydoc_attributes_as_param_list,
        "xref_param_type": app.config.numpydoc_xref_param_type,
        "xref_aliases": app.config.numpydoc_xref_aliases_complete,
        "xref_ignore": app.config.numpydoc_xref_ignore,
    }
    cfg.update(options or {})  # <- fails with "NoneType not callable"
```

The source of the problem is that because _AutoDocumenterOptions returns `None` for any missing attribute Python tries to call `options.items()` which in turn fails with the very confusing error :

```

Extension error (numpydoc.numpydoc)!

Versions
========

* Platform:         linux; (Linux-6.16.7-arch1-1-x86_64-with-glibc2.42)
* Python version:   3.15.0a0 (CPython)
* Sphinx version:   8.3.0
* Docutils version: 0.21.2
* Jinja2 version:   3.1.6
* Pygments version: 2.18.0

Last Messages
=============


    reading sources... [  3%]
    _tags/topic-tagging


    reading sources... [  4%]
    api/_afm_api

    Path to light image logo does not exist: _static/logo_light.svg
    Path to dark image logo does not exist: _static/logo_dark.svg

Loaded Extensions
=================

* sphinx.ext.mathjax (8.3.0)
* alabaster (1.0.0)
* sphinxcontrib.applehelp (2.0.0)
* sphinxcontrib.devhelp (2.0.0)
* sphinxcontrib.htmlhelp (2.1.0)
* sphinxcontrib.serializinghtml (2.0.0)
* sphinxcontrib.qthelp (2.0.0)
* sphinx.ext.autodoc.preserve_defaults (8.3.0)
* sphinx.ext.autodoc.type_comment (8.3.0)
* sphinx.ext.autodoc.typehints (8.3.0)
* sphinx.ext.autodoc (8.3.0)
* sphinx.ext.autosummary (8.3.0)
* sphinx.ext.graphviz (8.3.0)
* sphinx.ext.inheritance_diagram (8.3.0)
* sphinx.ext.intersphinx (8.3.0)
* sphinx.ext.ifconfig (8.3.0)
* IPython.sphinxext.ipython_console_highlighting (unknown version)
* IPython.sphinxext.ipython_directive (unknown version)
* numpydoc (1.10.0rc0.dev0)
* sphinx_gallery.gen_gallery (0.19.0)
* matplotlib.sphinxext.mathmpl (unknown version)
* matplotlib.sphinxext.plot_directive (3.11.0.dev1414+g5e621a4018.d20251010)
* matplotlib.sphinxext.roles (3.11.0.dev1414+g5e621a4018.d20251010)
* matplotlib.sphinxext.figmpl_directive (3.11.0.dev1414+g5e621a4018.d20251010)
* sphinxext.github (unknown version)
* sphinxext.math_symbol_table (unknown version)
* sphinxext.missing_references (unknown version)
* sphinxext.mock_gui_toolkits (unknown version)
* sphinxext.skip_deprecated (unknown version)
* sphinxext.redirect_from (unknown version)
* sphinx_copybutton (0.5.2)
* sphinx_design (0.6.0)
* sphinx_tags (0.4)
* sphinx.ext.linkcode (8.3.0)
* sphinxcontrib.video (unknown version)
* mpl_sphinx_theme (3.9.0)
* pydata_sphinx_theme (unknown version)

Traceback
=========

      File "/home/tcaswell/.virtualenvs/cp315/lib/python3.15/site-packages/sphinx/events.py", line 423, in emit
        raise ExtensionError(
        ...<4 lines>...
        ) from exc
    sphinx.errors.ExtensionError: Handler <function mangle_docstrings at 0x7f46477a8720> for event 'autodoc-process-docstring' threw an exception (exception: 'NoneType' object is not callable)


The full traceback has been saved in:
/tmp/sphinx-err-6obaxjdt.log

To report this error to the developers, please open an issue at <https://github.com/sphinx-doc/sphinx/issues/>. Thanks!
Please also report this if it was a user error, so that a better error message can be provided next time.
make: *** [Makefile:50: html] Error 2
```

### How to Reproduce

I hit this trying to build the Matplotlib docs, I apologize for not having a simpler reproducer. 

### Environment Information

```text
Platform:              linux; (Linux-6.16.7-arch1-1-x86_64-with-glibc2.42)
Python version:        3.15.0a0 (heads/main:03f5519d776, Aug 15 2025, 14:41:31) [GCC 15.2.1 20250813])
Python implementation: CPython
Sphinx version:        8.3.0  # master branch @ ad3f3cc062
Docutils version:      0.21.2
Jinja2 version:        3.1.6
Pygments version:      2.18.0
```

### Sphinx extensions

```python

```

### Additional context

_No response_

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -4,6 +4,10 @@ Release 9.0.1 (in development)
 Bugs fixed
 ----------
 
+* #13942: autodoc: Restore the mapping interface for options objects.
+  Patch by Adam Turner.
+* #13942: autodoc: Deprecate the mapping interface for options objects.
+  Patch by Adam Turner.
 
 Release 9.0.0 (released Nov 30, 2025)
 =====================================

--- a/doc/usage/extensions/autodoc.rst
+++ b/doc/usage/extensions/autodoc.rst
@@ -1393,9 +1393,8 @@ autodoc provides the following additional events:
    :param name: the fully qualified name of the object
    :param obj: the object itself
    :param options: the options given to the directive: an object with attributes
-      ``inherited_members``, ``undoc_members``, ``show_inheritance`` and
-      ``no-index`` that are true if the flag option of same name was given to the
-      auto directive
+      corresponding to the options used in the auto directive, e.g.
+      ``inherited_members``, ``undoc_members``, or ``show_inheritance``.
    :param lines: the lines of the docstring, see above
 
 .. event:: autodoc-before-process-signature (app, obj, bound_method)
@@ -1424,9 +1423,8 @@ autodoc provides the following additional events:
    :param name: the fully qualified name of the object
    :param obj: the object itself
    :param options: the options given to the directive: an object with attributes
-      ``inherited_members``, ``undoc_members``, ``show_inheritance`` and
-      ``no-index`` that are true if the flag option of same name was given to the
-      auto directive
+      corresponding to the options used in the auto directive, e.g.
+      ``inherited_members``, ``undoc_members``, or ``show_inheritance``.
    :param signature: function signature, as a string of the form
       ``'(parameter_1, parameter_2)'``, or ``None`` if introspection didn't
       succeed and signature wasn't specified in the directive.
@@ -1487,6 +1485,5 @@ member should be included in the documentation by using the following event:
    :param skip: a boolean indicating if autodoc will skip this member if the
       user handler does not override the decision
    :param options: the options given to the directive: an object with attributes
-      ``inherited_members``, ``undoc_members``, ``show_inheritance`` and
-      ``no-index`` that are true if the flag option of same name was given to the
-      auto directive
+      corresponding to the options used in the auto directive, e.g.
+      ``inherited_members``, ``undoc_members``, or ``show_inheritance``.

--- a/sphinx/ext/autodoc/_directive_options.py
+++ b/sphinx/ext/autodoc/_directive_options.py
@@ -1,14 +1,16 @@
 from __future__ import annotations
 
+import warnings
 from typing import TYPE_CHECKING
 
 from docutils.utils import assemble_option_dict
 
+from sphinx.deprecation import RemovedInSphinx11Warning
 from sphinx.ext.autodoc._sentinels import ALL, EMPTY, SUPPRESS
 from sphinx.locale import __
 
 if TYPE_CHECKING:
-    from collections.abc import Mapping, Set
+    from collections.abc import Iterable, Iterator, Mapping, Set
     from typing import Any, Final, Literal, Self
 
     from sphinx.ext.autodoc._property_types import _AutodocObjType
@@ -89,6 +91,99 @@ class _AutoDocumenterOptions:
     @classmethod
     def from_directive_options(cls, opts: Mapping[str, Any], /) -> Self:
         return cls(**{k.replace('-', '_'): v for k, v in opts.items() if v is not None})
+
+    # Mapping interface:
+
+    def __getitem__(self, item: str) -> Any:
+        warnings.warn(
+            'The mapping interface for autodoc options objects is deprecated, '
+            'and will be removed in Sphinx 11. Use attribute access instead.',
+            RemovedInSphinx11Warning,
+            stacklevel=2,
+        )
+        try:
+            return getattr(self, item)
+        except AttributeError:
+            raise KeyError(item) from None
+
+    def __setitem__(self, key: str, value: Any) -> None:
+        msg = f'{self.__class__.__name__!r} object does not support indexed assignment'
+        raise TypeError(msg)
+
+    def __delitem__(self, key: str) -> None:
+        msg = f'{self.__class__.__name__!r} object does not support indexed deletion'
+        raise TypeError(msg)
+
+    def __contains__(self, item: str) -> bool:
+        warnings.warn(
+            'The mapping interface for autodoc options objects is deprecated, '
+            'and will be removed in Sphinx 11. Use attribute access instead.',
+            RemovedInSphinx11Warning,
+            stacklevel=2,
+        )
+        return hasattr(self, item)
+
+    def __keys(self) -> list[str]:
+        return [key for key in dir(self) if not key.startswith('_')]
+
+    def __iter__(self) -> Iterator[str]:
+        warnings.warn(
+            'The mapping interface for autodoc options objects is deprecated, '
+            'and will be removed in Sphinx 11. Use attribute access instead.',
+            RemovedInSphinx11Warning,
+            stacklevel=2,
+        )
+        yield from self.__keys()
+
+    def __len__(self) -> int:
+        warnings.warn(
+            'The mapping interface for autodoc options objects is deprecated, '
+            'and will be removed in Sphinx 11. Use attribute access instead.',
+            RemovedInSphinx11Warning,
+            stacklevel=2,
+        )
+        return len(self.__keys())
+
+    def keys(self) -> Iterable[str]:
+        warnings.warn(
+            'The mapping interface for autodoc options objects is deprecated, '
+            'and will be removed in Sphinx 11. Use attribute access instead.',
+            RemovedInSphinx11Warning,
+            stacklevel=2,
+        )
+        yield from self.__keys()
+
+    def items(self) -> Iterable[tuple[str, Any]]:
+        warnings.warn(
+            'The mapping interface for autodoc options objects is deprecated, '
+            'and will be removed in Sphinx 11. Use attribute access instead.',
+            RemovedInSphinx11Warning,
+            stacklevel=2,
+        )
+        for key in self.__keys():
+            yield key, getattr(self, key)
+
+    def values(self) -> Iterable[Any]:
+        warnings.warn(
+            'The mapping interface for autodoc options objects is deprecated, '
+            'and will be removed in Sphinx 11. Use attribute access instead.',
+            RemovedInSphinx11Warning,
+            stacklevel=2,
+        )
+        for key in self.__keys():
+            yield getattr(self, key)
+
+    def get(self, key: str, default: Any | None = None) -> Any | None:
+        warnings.warn(
+            'The mapping interface for autodoc options objects is deprecated, '
+            'and will be removed in Sphinx 11. Use attribute access instead.',
+            RemovedInSphinx11Warning,
+            stacklevel=2,
+        )
+        try:
+            return getattr(self, key)
+        except AttributeError:
+            return default
 
 
 def identity(x: Any) -> Any:

